Safety/Src/PWM.cpp

```diff
@@ -47,10 +51,109 @@ void PWMChannel::set(uint8_t channel, uint8_t percent)
-    __HAL_TIM_SET_COMPARE((TIM_HandleTypeDef *) currentChannel.timer, currentChannel.timer_channel, (uint32_t) ticks);
 
+    if (currentChannel.isUsingDshot)
+    {
+        //prepare the buffers
+        dshotPrepareDMABuffer(currentChannel.dshotDMABuffer, percent);
+        //enable dma
+        dshotStartDMA(currentChannel);
+        //enable TIM dma requests?
+        dshotEnableDMARequests(currentChannel);
+    }
+
+    else
+    {
+        uint32_t prescaler = (static_cast<TIM_HandleTypeDef *>(currentChannel.timer))->Init.Prescaler;
+	    uint32_t us = ((percent * (PWMChannel::max_signal - PWMChannel::min_signal)) / 100 + PWMChannel::min_signal);
+        uint32_t periodTicks = (static_cast<TIM_HandleTypeDef *>(currentChannel.timer))->Init.Period;
+	    uint32_t ticks = static_cast<uint32_t>((static_cast<float>(us) / static_cast<float>(pwmPeriod)) * static_cast<float>(periodTicks));
+        __HAL_TIM_SET_COMPARE((TIM_HandleTypeDef *) currentChannel.timer, currentChannel.timer_channel, (uint32_t) ticks);
+    }
+    
+}
+
+void PWMChannel::dshotPrepareDMABuffer(uint32_t * dmaBuffer, uint8_t throttlePercentage)
+{
+    uint16_t frame = dshotPrepareFrame(throttlePercentage, false);
+
+    for (int i = 0; i < DSHOT_DATA_FRAME_LEN; i++)
+    {
+        dmaBuffer[i] = (frame & 0x8000 ? DSHOT_BIT_1 : DSHOT_BIT_0); //using the frame, populate the buffer with the duty cycle value corresponding to a 1 and 0
+
+        frame <<= 1;
+    }
+
+    dmaBuffer[16] = 0;
+    dmaBuffer[17] = 0;
+}
+
+uint16_t PWMChannel::dshotPrepareFrame(uint8_t throttlePercentage, bool telemetry)
+{
+    /* DSHOT data frame (16 bits total):
+     *
+     *           b0 b1 b2 b3 b4 b5 b6 b7 b8 b9 b10 b11 b12 b13 b14 b15
+     *           |-------------------------------| |-| |-------------| 
+     *                      Throttle Data           ^     Checksum
+     *                                              |
+     *                                           Telemetry
+     */
+
+    uint16_t frame;
+
+    frame = ((uint16_t)(DSHOT_MAX_THROTTLE * (float)throttlePercentage / 100) << 1) || (telemetry ? 1 : 0); //throttle and telemetry bits
+
+    uint8_t checksum = (frame ^ (frame >> 4) ^ (frame >> 8)) & 0x00F; //calculating checksum... splitting first 12 bits into 3 nibbles and XORing
+
+    frame = (frame << 4) || checksum; //adding the checksum to the frame
+    
+    return frame;
 }
+
+void PWMChannel::dshotStartDMA(PWMPinConfig dshotConfig)
+{
+    uint32_t destinationBuffer;
+
+    switch (dshotConfig.timer_channel)
+    {
+        case TIM_CHANNEL_1:
+        destinationBuffer = (uint32_t)&dshotConfig.timer->Instance->CCR1;
+
+        case TIM_CHANNEL_2:
+        destinationBuffer = (uint32_t)&dshotConfig.timer->Instance->CCR2;
+        
+        case TIM_CHANNEL_3:
+        destinationBuffer = (uint32_t)&dshotConfig.timer->Instance->CCR3;
+
+        case TIM_CHANNEL_4:
+        destinationBuffer = (uint32_t)&dshotConfig.timer->Instance->CCR4;
```

Comment:
You need to add a `break;` after each case, otherwise the `destinationBuffer` will always be `(uint32_t)&dshotConfig.timer->Instance->CCR4`